x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -0,0 +1,185 @@
+    }
+
+    void setMaxFailureRetries(int value) {
+        this.maxFailureRetries = value;
+    }
+
+    public BulkResponse indexWithRetry(String jobId,
+                                       String indexName,
+                                       ToXContent object,
+                                       ToXContent.Params params,
+                                       WriteRequest.RefreshPolicy refreshPolicy,
+                                       String id,
+                                       Supplier<Boolean> shouldRetry,
+                                       Consumer<String> msgHandler) throws IOException {
+        BulkRequest bulkRequest = new BulkRequest().setRefreshPolicy(refreshPolicy);
+        try (XContentBuilder content = object.toXContent(XContentFactory.jsonBuilder(), params)) {
+            bulkRequest.add(new IndexRequest(indexName).id(id).source(content));
+        }
+        return bulkIndexWithRetry(bulkRequest, jobId, shouldRetry, msgHandler);
+    }
+
+    public BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest,
+                                           String jobId,
+                                           Supplier<Boolean> shouldRetry,
+                                           Consumer<String> msgHandler) {
+        return bulkIndexWithRetry(bulkRequest, jobId, shouldRetry, msgHandler, 10);
+    }
+
+    BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest,
+                                    String jobId,
+                                    Supplier<Boolean> shouldRetry,
+                                    Consumer<String> msgHandler,
+                                    int msgBarrier) {
+        int currentMin = MIN_RETRY_SLEEP_MILLIS;
+        int currentMax = MIN_RETRY_SLEEP_MILLIS;
+        int currentAttempt = 0;
+        BulkResponse bulkResponse = null;
+        while(currentAttempt <= maxFailureRetries) {
+            bulkResponse = bulkIndex(bulkRequest);
+            if (bulkResponse.hasFailures() == false) {
+                return bulkResponse;
+            }
+            if (shouldRetry.get() == false) {
+                throw new ElasticsearchException("[{}] failed to index all results. {}", jobId, bulkResponse.buildFailureMessage());
+            }
+            if (currentAttempt > maxFailureRetries) {
+                LOGGER.warn("[{}] failed to index after [{}] attempts. Setting [xpack.ml.persist_results_max_retries] was reduced",
+                    jobId,
+                    currentAttempt);
+                throw new ElasticsearchException("[{}] failed to index all results after [{}] attempts. {}",
+                    jobId,
+                    currentAttempt,
+                    bulkResponse.buildFailureMessage());
+            }
+            currentAttempt++;
+            // Before 10, most retries are fairly fast,
+            // After 10, the time between quickly gets on the magnitude of minutes
+            if (currentAttempt > msgBarrier) {
```

Comment:
IMHO: failing is not expected, I would not consider it logspam to log/audit every time.

Reply:
I am logging and sending a message to the handler every time now